NitroxClient/GameLogic/Bases/BuildingHandler.cs

```diff
@@ -0,0 +1,372 @@
+            {
+                continue;
+            }
+            Log.Debug($"[{pieceDeconstructed.OperationId}] Found a BaseDeconstructable {baseDeconstructable.name}, will now deconstruct it manually");
+            using (PacketSuppressor<BaseDeconstructed>.Suppress())
+            using (PacketSuppressor<PieceDeconstructed>.Suppress())
+            using (PacketSuppressor<WaterParkDeconstructed>.Suppress())
+            {
+                Temp.Fill(pieceDeconstructed);
+                baseDeconstructable.Deconstruct();
+                Temp.Reset();
+            }
+            tracker.RegisterOperation(pieceDeconstructed.OperationId);
+            BasesCooldown[pieceDeconstructed.BaseId] = DateTimeOffset.Now;
+            yield break;
+        }
+        Log.Error("Couldn't find BaseDeconstructable to be destructed");
+        tracker.FailedOperations++;
+    }
+
+    public void AskForResync()
+    {
+        if (!Multiplayer.Main || !Multiplayer.Main.InitialSyncCompleted)
+        {
+            return;
+        }
+        TimeSpan deltaTime = DateTimeOffset.Now - LatestResyncRequestTimeOffset;
+        if (deltaTime.TotalMilliseconds < RESYNC_REQUEST_COOLDOWN)
+        {
+            double timeLeft = RESYNC_REQUEST_COOLDOWN * 0.001 - deltaTime.TotalSeconds;
+            Log.InGame(Language.main.Get("Nitrox_ResyncOnCooldown").Replace("{TIME_LEFT}", timeLeft.ToString()));
+            return;
+        }
+        LatestResyncRequestTimeOffset = DateTimeOffset.Now;
+
+        this.Resolve<IPacketSender>().Send(new BuildingResyncRequest());
+        Log.InGame(Language.main.Get("Nitrox_ResyncRequested"));
+    }
+
+    public void StartResync(Dictionary<Entity, int> entities)
+    {
+        Resyncing = true;
+        FailedOperations = 0;
+        BuildQueue.Clear();
+        InitializeOperations(entities.ToDictionary(pair => pair.Key.Id, pair => pair.Value));
+    }
+
+    public void StopResync()
+    {
+        Resyncing = false;
+    }
+
+    public void InitializeOperations(Dictionary<NitroxId, int> operations)
+    {
+        foreach (KeyValuePair<NitroxId, int> pair in operations)
+        {
+            Log.Debug($"Resetting {pair.Key} to {pair.Value}");
+            EnsureTracker(pair.Key).ResetToId(pair.Value);
+        }
+    }
+
+
+    public static Transform GetParentOrGlobalRoot(NitroxId id)
+    {
+        if (id != null && NitroxEntity.TryGetObjectFrom(id, out GameObject parentObject))
+        {
+            return parentObject.transform;
+        }
+        else
+        {
+            return LargeWorldStreamer.main.globalRoot.transform;
+        }
+    }
+
+    private void CleanCooldowns()
+    {
+        DateTimeOffset now = DateTimeOffset.Now;
+        List<NitroxId> keysToRemove = BasesCooldown.Where(entry => (now - entry.Value).TotalMilliseconds >= MULTIPLAYER_BUILD_COOLDOWN).Select(e => e.Key).ToList();
+        keysToRemove.ForEach(key => BasesCooldown.Remove(key));
```

Comment:
This won't work because you can't modify the current collection (which is done when using ForEach's remove)

Reply:
Isn't `Where()` creating a new enumerator/list?

Reply:
I'm unsure of how it really works but I've tested your proposition and it didn't work for the reason mentionned above.
![image](https://github.com/SubnauticaNitrox/Nitrox/assets/24827220/980979e7-e289-4877-8e95-2d2912d6b743)

Reply:
Some improvement here is warranted as it's running every frame. Unfortunately `Dictionary` allocates ever iteration because of the `Enumerator` but I could reduce allocations to 17% of original using a pooled array:
```cs
public static void RemoveWhere<TKey, TValue, TParameter>(this IDictionary<TKey, TValue> dictionary, TParameter extraParameter, Func<TValue, TParameter, bool> predicate)
{
    int toRemoveIndex = 0;
    TKey[] toRemove = ArrayPool<TKey>.Shared.Rent(dictionary.Count);
    try
    {
        foreach (KeyValuePair<TKey, TValue> item in dictionary)
        {
            if (predicate.Invoke(item.Value, extraParameter))
            {
                toRemove[toRemoveIndex++] = item.Key;
            }
        }
        for (int i = 0; i < toRemoveIndex; i++)
        {
            dictionary.Remove(toRemove[i]);
        }
    }
    finally
    {
        ArrayPool<TKey>.Shared.Return(toRemove, true);
    }
}
```

[Benchmark code with results](https://gist.github.com/Measurity/a1130e46cbf805fb79258603431a19d8)